druid-shell/src/platform/mac/window.rs

```diff
@@ -866,31 +884,55 @@ impl WindowHandle {
         }
     }
 
+    /// Request a new paint, but without invalidating anything.
     pub fn request_anim_frame(&self) {
+        let view = self.nsview.load();
         unsafe {
-            // TODO: synchronize with screen refresh rate using CVDisplayLink instead.
-            let () = msg_send![*self.nsview.load(), performSelectorOnMainThread: sel!(redraw)
-                withObject: nil waitUntilDone: NO];
+            if let Some(view) = (*view).as_ref() {
+                let view_state: *mut c_void = *view.get_ivar("viewState");
+                (*(view_state as *mut ViewState)).display_link.start();
+            }
         }
     }
 
-    // Request invalidation of the entire window contents.
+    /// Request invalidation of the entire window contents.
     pub fn invalidate(&self) {
+        let view = self.nsview.load();
+        let bounds: NSRect = unsafe { msg_send![*view, bounds] };
+        let rect = Rect::from_origin_size(
+            (bounds.origin.x, bounds.origin.y),
+            (bounds.size.width, bounds.size.height),
+        );
         unsafe {
-            // We could share impl with redraw, but we'd need to deal with nil.
-            let () = msg_send![*self.nsview.load(), setNeedsDisplay: YES];
+            if let Some(view) = (*view).as_ref() {
+                let view_state: *mut c_void = *view.get_ivar("viewState");
+                (*(view_state as *mut ViewState))
+                    .invalid_region
+                    .lock()
+                    .unwrap()
+                    .add_rect(rect);
+            }
+            let () = msg_send![*view, setNeedsDisplayInRect: bounds];
```

Comment:
Can't you just do `setNeedsDisplay` here instead of calculating the bounds yourself?

Reply:
Yes that would have the same effect, but the rect would still need to get added to `invalid_region`, so `draw_rect` would repaint the whole screen.